### PR TITLE
feat: enable streaming weight updates

### PR DESCRIPTION
**Pull request by hchings**
## Description

This MR rewrites PyTorch backend's `_load_weights_impl` (which currently is a brute-force O(N^2) of looping through all `module.named_parameters` and for each module it loops through full weights) into a more efficient partial weight updates for further RL integration. As an example, partial weight_loading of other inference FW is called [here](https://github.com/NVIDIA/NeMo-RL/blob/main/nemo_rl/models/generation/vllm_backend.py#L58) in nemo-rl.

This is part of [TRTLLM-4942](https://jirasw.nvidia.com/browse/), where majority of the works is still blocked by:
1. Migration to Ray orchestration (if such direction is approved) and how will we do collective_rpc calls.
2. Virtual memory allocator for waking up partial weights, which requires TRTLLM C++ to add support for sleep and wakeup [[doc](https://docs.google.com/document/d/1lyIwcZ0CxhSiqeHJN6zG5WtDFGInWzVdgf4T7YGTabM/edit?tab=t.), [MR](https://github.com/NVIDIA/TensorRT-LLM/pull/5034/files) by Yuan].

Based on brief discussion w/ nemo-rl, a training step has `2-3` `load_weight()` calls for a 70B model with nemo-rl's bucketing approach.  I need to first check and confirm whether improving time of load_weight() itself is even worth for nemo-rl, given the weight loading overhead is likely to be in somewhere else like the rpc call. 

TODOs:
- [x] wrap up impl
- [ ] Understand perf implication in Nemo-RL (WIP)
- [ ] migrate all models under `_torch/models` w/ customized `load_weights()` to use new method 
    - exclude `deepseekv3`, `qwen_moe`, `qwen3_moe`
    - not sure yet: `mamba_hybrid`
- [ ] testing
    - accuracy 
    - perf - right now I'm only seeing ~8% improvements for a 70B model single GPU with this draft (which can be futhered improved) if I load partial weights. So unless `load_weight()` is called a huge number of times in common RLHF workloads, it might not be worthy to improve `load_weight()` itself.

## Test Coverage

Currently using `examples/pytorch/quickstart.py` with various models to test locally. 

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Comment by hchings** [comment]
Closing as latest discussion is to focus on IPC for TRTLLM/MPI + VeRL tentative integration. 

After Ray migration is more mature, I'll revisit this.
